### PR TITLE
Refactor EndpointStrategy to support a pick method that takes timeout.

### PR DESCRIPTION
**Pull request by chtyim**
- Removed the inferior TimeEndpointStrategy in favor for the new pick method

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java [hunk 1]
```diff
@@ -61,4 +68,20 @@ public URL getServiceURL(String serviceId) {
    * @return the {@link DiscoveryServiceClient} for Service Discovery
    */
   protected abstract DiscoveryServiceClient getDiscoveryServiceClient();
+
+  private URL createURL(@Nullable Discoverable discoverable, String applicationId, String serviceId) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java]: Should we add ``@Nullable`` for the method as well, although it is private?

File: cdap-common/src/main/java/co/cask/cdap/common/discovery/EndpointStrategy.java [hunk 2]
```diff
@@ -24,7 +26,17 @@
 
   /**
    * Picks a {@link Discoverable} using its strategy.
-   * @return A {@link Discoverable} based on the stragegy or {@code null} if no endpoint can be found.
+   * @return A {@link Discoverable} based on the strategy or {@code null} if no endpoint can be found.
    */
   Discoverable pick();
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/discovery/EndpointStrategy.java]: Should all instances of both these methods have ``@Nullable`` annotation?

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java [hunk 3]
```diff
@@ -103,8 +112,23 @@ public URL getServiceURL() {
   @Override
   public URL getServiceURL(long timeout, TimeUnit timeoutUnit) {
     String discoveryName = String.format("service.%s.%s.%s", accountId, applicationId, serviceName);
-    String path = String.format("%s/namespaces/%s/apps/%s/services/%s/methods/",
+    ServiceDiscovered discovered = discoveryServiceClient.discover(discoveryName);
+    return createURL(new RandomEndpointStrategy(discovered).pick(timeout, timeoutUnit), applicationId, serviceName);
+  }
+
+  private URL createURL(@Nullable Discoverable discoverable, String applicationId, String serviceName) {
```
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java]: ``@Nullable`` here too.

Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java]: Also, is there a utility class where we could share this ``createURL`` method between main code and test code, although I guess the point of doing this refactor was also to remove the static utility class ``ServiceDiscoveries``? Not sure if it would be a good idea to keep that class with just this method?

**Comment by bdmogal** [comment]
Minor comments. Overall, the refactor LGTM.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java (hunk 3 above)
Comment (chtyim) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java]: I am not a big fan of util class. Ideally they can be extends from some abstract base class. However it requires bigger refactor on how the classes are organized. Probably do it in a separate PR as it's not the focus of this PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java]: added

File: cdap-common/src/main/java/co/cask/cdap/common/discovery/EndpointStrategy.java (hunk 2 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/discovery/EndpointStrategy.java]: added

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java (hunk 3 above)
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultServiceManager.java]: Sounds good! Don't like utility classes either.

**Comment by bdmogal** [comment]
@chtyim Just one question about ``@Nullable``:

Do we need to add the annotation to implementing classes if the annotation is present on the interface methods? If so, you may want to add it to ``pick`` methods in ``RandomEndpointStrategy`` and ``AbstractEndpointStrategy``. Sorry for missing it last time :)

Other than that, LGTM.

**Comment by chtyim** [comment]
no, as that's stated in the interface

**Comment by bdmogal** [comment]
Cool, :+1:
